Added intro to linked list tutorial

JacobGa03 opened the pull request:
Added "Linked List Basics" tutorial to the linked list section.

JacobGa03 committed:
Added intro to linked list tutorial

Added "Linked List Basics" tutorial to the linked list section.

jerrettl commented on src/cs1/index.md:
```diff
@@ -11,4 +11,6 @@ toc: false
 
 ## Linked Lists
 
+[Linked List Basic](linked-lists/linked)
```
Updating to match file name and title:

```suggestion
[Linked List Basics](linked-lists/linked-list-basics.md)
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+A Linked List is exactly what it sounds like, a list where all of the elements are linked together like a chain.
+
+Here is an example of a Linked List
```
```suggestion
Here is an example of a Linked List:
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Since we have access constant access to the head,  we don't have to do any traversing. This make inserting in the front an `O(1)` operation.
+
+Now let's look at the code
```
```suggestion
Now let's look at the code:
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++-----+    +------+
+```                  
+
+Since we have access constant access to the head,  we don't have to do any traversing. This make inserting in the front an `O(1)` operation.
````
```suggestion
Since we have constant-time access to the head, we don't have to do any traversing. This makes inserting in the front an `O(1)` operation.
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++----------+
+```                  
+
+We then iterate tempHeap until the node we are at has `NULL` next node. Then, we can add the new node to our list. 
````
I may tweak this wording here on the first sentence, maybe something along these lines to clarify what the role of `NULL` is:

```suggestion
We then iterate `tempHead` until the node we are at has `NULL` as the value of the `next` node. Then, we can add the new node to our list. 
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
+}node;
+```
+
+Here we don't store the whole list in a data structure when compared to the array; rather we store the individual nodes that compose the list itself.
````
I notice in this section there is no explicit mention of it, but it is used later as an assumption: The end of a linked list is indicated where the `next` pointer is `NULL`, so just a small note here may be good.

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Here we insert `7` to the end of the list with our first method.
+
+We will create a temporary `node*` that will allow us to find the end of the list. Consider what happens if we do not create a temporary pointer. If we use the head to traverse the rest of the list then we will loose the nodes we pass over. This lose of unfreed memory is called a `memory leak`.
```
Small wording tweak, up to your consideration:

```suggestion
We will create a temporary `node*` that will allow us to find the end of the list. Consider what happens if we do not create a temporary pointer. If we use the `head` variable itself to traverse the rest of the list then we will lose the nodes we pass over. This loss of unfreed memory is called a `memory leak`.
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Another way we can insert to back is by having another pointer to the end of the list. Typically we call this node `tail`. Keeping track of `tail` allows us to insert in `O(1)` time.
+
+Let's look at the `O(n)` insert code
```
```suggestion
Let's look at the `O(n)` insert code:
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
+              +-----+
+```
+
+Once we've identified the last node in the list we simply add our new node to the back of the list. Since we have to traverse the WHOLE list to find the end, this makes this approach take `O(n)` time.
````
Over here, I would specify exactly what changes would be made, so pointing the last node to your new node, and pointing the new node to `NULL`.

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++----------+       +------+
+```
+
+We set `13` next value as `15` and then we set `tempHead`'s next value to 13. Now we've inserted `13` in the correct place within the list
````
This is one suggestion, but something similar to this as you think is good!

```suggestion
We set `13`'s next value as `15` and then we set `tempHead`'s next value to `13`. Now we've inserted `13` in the correct place within the list.
```

Technically this also has the same comment from the other PR about distinguishing between a node and the value within it, so it may be worth describing this at least once somewhere in the page, but since it was already mentioned I won't harp on it too much here.

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++------+    +------+    +------+    +------+    +------+
+```
+
+What we will need to do is traverse the list until we find the first node that is greater than the value we want to add into our list. Just like add to back, we will need a temporary node in order to traverse over the list.
````
Just more semantics, but in this case for tracking, only a pointer is made instead of a node, so I would make that clear.

```suggestion
What we will need to do is traverse the list until we find the first node that is greater than the value we want to add into our list. Just like add to back, we will need a temporary pointer in order to traverse over the list.
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Once we "patch over" the deleted node, then we can free it.
+
+Let's delete `15` from the list
```
```suggestion
Let's delete `15` from the list:
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++------+    +------+    +------+    +------+    +------+
+```
+
+Just like before, we will need a temporary node to traverse the rest of the list. Once we find the node RIGHT BEFORE `15`, then we can perform the "patch over" and deletion of `15`.
+
+```
++------+    +------+    +------+    +------+
+|  11  | -> |  15  | -> |  18  | -> | NULL |
++------+    +------+    +------+    +------+
+   ^
+   |
++----------+
+| tempHead |
++----------+
+```
+
+Once we are here we will save `15` in another temporary node, then "patch over", then free `15.
````
Same as earlier, should this be a temporary node or a temporary pointer? Since we are working with these new concepts in linked lists, that distinction can be very important! (From experience as a TA you can never be too careful with thinking that a student will assume to know what you mean.)

JacobGa03 replied:
Perhaps for `NULL` we can just use an `X` like I used for the binary trees?

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
````diff
@@ -0,0 +1,401 @@
++----------+
+```
+
+Once we are here we will save `15` in another temporary node, then "patch over", then free `15.
````
```suggestion
Once we are here we will save `15` in another temporary node, then "patch over", then free `15`.
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Since at WORST we might have to traverse the whole list, this makes delete an `O(n)` operation.
+
+Now let's look at the delete code
```
```suggestion
Now let's look at the delete code:
```

jerrettl commented on src/cs1/linked-lists/linked-list-basics.md:
```diff
@@ -0,0 +1,401 @@
+
+Searching in a Linked List is probably the easiest operation to perform since it is just a Linked List traversal.
+
+All we need is, you guessed it, a temporary node so we can search the whole list for the value what we are looking for. Once we find the value we are looking for we can return `1` to indicate a successful search or `0` for an unsuccessful.
```
```suggestion
All we need is, you guessed it, a temporary pointer so we can search the whole list for the value what we are looking for. Once we find the value we are looking for we can return `1` to indicate a successful search or `0` for an unsuccessful.
```

jerrettl reviewed (commented):
Okay I think I am done getting through everything whew! As before, this covers a really nice breadth of a lot of details and information!

All I have are more little details here and there to make it all as clear as possible for the reader. I remember back when seeing students in office hours and how important it was to get all the little terms and precision, especially when dealing with pointers. (AVL self-balancing algorithms were also a doozy in this regard.) There are other spots, but hopefully that gives a general idea of what I am thinking. Hopefully this wasn't too crazy, so same thing, comments from #62 and #67 apply here, and just real minor stuff throughout.

- This is completely a subjective thing, but for the diagrams I see everything is surrounded by a box. This is good for visuals, but I worry this may be a little confusing as it appears like everything is a node, when in some cases they aren't actually nodes. Like `NULL` has a box around it, but `NULL` doesn't represent a node like the others. Instead, it is just a value set on another node's `->next` member. It may be worth considering whether simple pointer variables like `tempHead` get the same boxing as a full node structure as well, since this kind of looks like it's an extra node in the chain altogether.

- For readability, when the arrows start getting more complicated towards the end, you may consider using these [Unicode box drawing characters](https://www.w3.org/TR/xml-entity-names/025.html). They're a bit of a pain in the neck to use, but can come out looking really cool!

JacobGa03 committed:
Update src/cs1/index.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Formating Fix/Spelling Errors

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Update src/cs1/linked-lists/linked-list-basics.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

JacobGa03 committed:
Fixed Up Linked List Pictures

Fixed up some of the `text` descriptions of the different algorithms. Made it so `NULL` is now `X`. It is also stated that the end of the list points to null.
